druntime/src/core/math.d

```diff
@@ -267,3 +269,1299 @@ T toPrec(T:real)(real f)  { pragma(inline, false); return f; }
+                    // magnitude of the returned low double. It may even be a
+                    // denormal.
+                    explo -= ex;
+                    ixl &= 0x000f_ffff_ffff_ffffUL;
+                    y.vul[1] &= 0x8000_0000_0000_0000UL;
+                    if (explo <= 0)
+                    {
+                        // Handle denormal low double.
+                        if (explo > -52)
+                        {
+                            ixl |= 1L << 52;
+                            ixl >>= 1 - explo;
+                        }
+                        else
+                        {
+                            ixl = 0;
+                            y.vul[1] = 0;
+                            if ((y.vul[0] & 0x7ff0_0000_0000_0000UL) == (1023L << 52))
+                            {
+                                // This can happen if the input was something
+                                // like 0x1p1000 - 0x1p-1000.
+                                y.vul[0] -= 1L << 52;
+                                ex += 1;
+                            }
+                        }
+                        explo = 0;
+                    }
+                    y.vul[1] |= (explo << 52) | ixl;
+                }
+                exp = ex;
+                return y.rv;
+            }
+            else
+                static assert(false, "frexp() not implemented");
+        }
+
+        // Params:
+        //      x = input value
+        // Returns:
+        //      1 if sign bit of x is set, 0 if not
+        pragma(inline, true)
+        static int signbit_inline(real x)
+        {
+            if (__ctfe)
+            {
+                // Precision can change, but sign won't change at CTFE.
+                double dx = cast(double)x;
+                return 0 > *cast(long*)&dx;
+            }
+            // Get the index of the sign when represented as a ubyte array.
+            version (LittleEndian)
+            {
+                static if (real.mant_dig == 64 || real.mant_dig == 53)
+                    enum pos = 9;
+                else static if (real.mant_dig == 113)
+                    enum pos = 15;
+                else static if (real.mant_dig == 106)
+                    enum pos = 7;
+                else
+                    static assert(false, "signbit() not implemented");
+            }
+            else
+                enum pos = 0;
+            return ((cast(ubyte *)&x)[pos] & 0x80) != 0;
+        }
+
+        // Params:
+        //      x = input value
+        // Returns:
+        //      a multiple of 1/32 that is within 1/32 of x
+        pragma(inline, true)
+        static real reduc_inline(real x)
+        {
+            return floor_inline(x * 32) / 32;
+        }
+
+        // Evaluate polynomial
+        // Params:
+        //      x = the value to evaluate
+        //      A = array of coefficients
+        pragma(inline, true)
+        static real poly_inline(alias A)(real x)
```

Comment:
All of these functions could probably be put inside a template somewhere else so there's no risk of having a bunch of them in the binary N times over.

Reply:
Note: all these `_inline` functions do not handle non-finite (and in some cases subnormal) input values because those are assumed to be taken care of by the "special cases" handling of outer `pow()` function.